sqlite.lisp

```diff
@@ -0,0 +1,122 @@
+(in-package #:cl)
+
+(defpackage #:scheduler-sqlite
+  (:use #:cl #:alexandria #:sqlite #:iterate)
+  (:import-from #:scheduler
+                #:sqlite-scheduler)
+  (:import-from #:scheduler-implementation
+                #:scheduler #:in-memory-scheduler 
+                #:start-scheduler #:stop-scheduler
+                #:create-scheduler-task
+                #:read-scheduler-task
+                #:update-scheduler-task
+                #:delete-scheduler-task
+                #:list-scheduler-tasks
+                ;; entry
+                #:scheduler-entry
+                #:cron-entry
+                #:parse-entry
+                #:parse-cron-entry
+                ;; task
+                #:task
+                #:task-time-specs
+                #:task-command
+                #:task-last-execution
+                #:task-next-execution
+                #:task-source-entry
+                #:execute-task)
+  (:export #:sqlite-scheduler
+           ;; task
+           #:sqlite-task))
+
+(in-package #:scheduler-sqlite)
+
+(defvar *db-file-path*
+  "Path to the SQLite DB file.")
+
+(defclass sqlite-scheduler (scheduler)
+  ())
+
+(defmethod initialize-instance :after ((scheduler sqlite-scheduler) &key db-path)
+  (assert db-path)
+  (setf *db-file-path* db-path)
+  (sqlite:with-open-database (db *db-file-path*)
+    (sqlite:execute-non-query
+     db "create table if not exists tasks (name text primary key, task text not null)")))
+
+(defclass sqlite-task (task)
+  ((name :initarg :name :accessor task-name)))
```

Comment:
Having a non-standard obligatory accessor and initarg prevents using the database as a drop-in replacement for the the `in-memory-scheduler`.

I think that the scheduler when started should read all entries into memory and operate on task instances, and when something changes it should write-through. Tasks should have autogenerated id, i.e

```
create table tasks (id integer primary key autoincrement, …)
```

Reply:
You need a static reference to each task to be able to delete or change it without analysing each task in the list. I tried auto-generated IDs initially, each time you change the time spec of the command details it creates another record.

Without names you somehow need to determine the specific task's id by analysis, which would require additional application code each time you want to make a timespec change. It will be like migrations for the scheduler.

Names was the least intrusive method I could come up with. I am open for other ideas.

Reply:
What I'm suggesting is to have a class `sqlite-task` that has the slot `id`, so when all tasks are read, then the `id` is also available to the lisp process. The point is that the user doesn't need to know about such slot at all.

Reply:
The user needs to now the task's "name" / id because it is necessary to change a task's definition without doing a lot of custom coding. When the timespec changes you want to make the change, recompile and run the app. No migrations or other things.

How would such a situation be handled with the opaque ID you suggest? Am I missing something?

Reply:
When you update the task, then you supply the task object as the second argument, so you may source the `id` from that object.

The main question is how to "find" the object after a cold reboot, when all tasks are re-read from the database (and the user doesn't have the old "handle") -- imho if you want to edit a particular command then you should find the task you want to edit and pass it to the ~update-scheduler-task~ (instead of trying to find the "matching" task in ~create-scheduler-task~).

It is the database what stores task definitions, not the source code. That's the point of persistence.

i.e
```
(in-package clim-user)

(define-presentation-method present
    (object (type scheduler-implementation:task) stream view &key)
  (declare (ignore type view))
  (format stream "~a" (scheduler-implementation:task-source-entry object)))

(defun list-tasks (frame stream)
  (surrounding-output-with-border (stream :shape :underline)
                                  (format stream "Tasks~%~%"))
  (terpri stream)
  (format-textual-list
   (scheduler:list-scheduler-tasks (scheduler frame))
   (lambda (object stream)
     (present object 'scheduler-implementation:task :stream stream))
   :separator #\newline)

  (terpri stream)
  (terpri stream)
  (surrounding-output-with-border (stream :shape :underline)
                                  (format stream "Example strings~%"))
  (terpri stream)
  (format-textual-list
   (list "H/3 * * * * (scheduler-implementation::%foobar)"
         "@reboot (scheduler-implementation::%xxx)")
   (lambda (object stream)
     (present object 'string :stream stream))
   :separator #\newline))

(define-application-frame scheduler-manager ()
  ((scheduler :initform (make-instance 'scheduler:in-memory-scheduler)
              :initarg :scheduler
              :reader scheduler))
  (:geometry :width 800 :height 600)
  (:pane :application :display-function 'list-tasks))

(define-scheduler-manager-command com-create
    ((time-entry 'string :gesture :select))
  (with-application-frame (frame)
    (scheduler:create-scheduler-task
     (scheduler frame) time-entry)))

(define-scheduler-manager-command com-edit
    ((task 'scheduler-implementation:task))
  (terpri)
  (terpri)
  (with-application-frame (frame)
    (let ((new-entry (accept 'string :prompt "New string")))
      (scheduler:update-scheduler-task (scheduler frame) task
                                       :cron-entry new-entry))))

(define-scheduler-manager-command com-delete
    ((task 'scheduler-implementation:task ))
  (terpri)
  (terpri)
  (with-application-frame (frame)
    (scheduler:delete-scheduler-task (scheduler frame) task)))

(define-presentation-to-command-translator tr-edit
    (scheduler-implementation:task com-edit scheduler-manager :gesture nil)
  (object)
  `(,object))

(define-presentation-to-command-translator tr-delete
    (scheduler-implementation:task com-delete scheduler-manager :gesture nil)
  (object)
  `(,object))

(find-application-frame 'scheduler-manager)
```

Reply:
I think we have different use cases in mind. I don't use the sqlite persistence for storing auto-generated schedules. My tasks are statically and fully defined in the source code. The sqlite backing is needed to store the last and next execution times reliably.

Tasks perform potentially dangerous actions like sending emails to the complete client base. It is not acceptable to perform the action twice because I restarted the application. 

This is also why I need a name or other static, hard coded, identifier in the source code. I sometimes change a particular task's time-spec or the command-string. This does not make it a different task, it is the same task that must be updated when the application restarts, without having to think about migrations or code to locate the task that needs to be updated.

Incidentally, there is currently a bug which prevents me from changing an existing task's time-spec. Fixing that was the reason I started working on the PR again.

Reply:
yes, I've noticed the bug when I was writing this clim application :)

I'd like to have schedulers be interchangeable and orthogonal to tasks whenever possible. I can see two paths forward:

1. we could bring _optional_ name  designators to the class `task` (and that would constitute a reasonable feature)

2. sqlite backend could handle task names separately without subclassing the task class - by specializing relevant methods on the class symbol

Here is a pseudocode that illustrates the second approach:
```
(defmethod create-scheduler-task ((scheduler sqlite-scheduler) entry
                                  &key name ...)
  ;; The name should be optional to allow storing anonymous entries.
  (when (and name (find-in-sqlite name))
    (error "Can't overwrite the task"))
  (insert-with-name task (call-next-method)))

(defmethod read-scheduler-task ((scheduler sqlite-scheduler) (task symbol))
  (find-in-sqlite task))

(defmethod update-scheduler-task ((scheduler sqlite-scheduler) (task symbol) ...)
  (unless (find-in-sqlite task)
    (error "Can't update task that does not exist"))
  (update-with-name task (find-in-sqlite task) ...))

(defmethod delete-scheduler-task ((scheduler sqlite-scheduler) (task symbol))
  (unless (find-in-sqlite task)
    (error "Can't delete the task that does not exist"))
  (delete-with-name task))

;;; redefinition semantics can be handled client-side, i.e with a macro like this:
(defmacro define-named-task (a-scheduler task-name ...)
  `(aif (find-in-sqlite ,task-name)
        (update-scheduler-task a-scheduler ^it        ...)
        (create-scheduler-task a-scheduler ,task-name ...)))
```

the important part is to have the name optional, in other words, the auto-incremented `ID` should be assigned to each task regardless, and the column `NAME` type should be `(or string null)`.